net: Add CNodeOptions and increase constness

ajtowns opened the pull request:
Adds CNodeOptions to make it easier to add optional parameters to the CNode constructor, and makes prefer_evict and m_permissionFlags actually const.

ajtowns committed:
net: add CNodeOptions for optional CNode constructor params

MarcoFalke commented on src/net.h:
```diff
@@ -344,7 +351,7 @@ class CNode
     const std::unique_ptr<const TransportSerializer> m_serializer;
 
-    NetPermissionFlags m_permissionFlags{NetPermissionFlags::None}; // treated as const outside of fuzz tester
+    const NetPermissionFlags m_permissionFlags;
```
nit (feel free to ignore), but it always felt weird to me to mix camels with snakes. Suggested commit:


```
    scripted-diff: Rename m_permissionFlags to m_permission_flags
    
    Seems better to use consistent style within one symbol name.
    Also, rename permissionFlags to permission_flags.
    
    -BEGIN VERIFY SCRIPT-
     sed -i 's/permissionFlags/permission_flags/g' $(git grep -l permissionFlags)
    -END VERIFY SCRIPT-

ajtowns replied:
Camels eaten by snakes.

vasild reviewed (commented):
Concept ACK

Why not put all arguments in `CNodeOptions`? It seems a bit confusing to have some arguments passed directly and some via `CNodeOptions`. Also, if both are used, then it is unclear which way should be used by a newly added argument in the future.

hebasto commented:
Concept ACK.

ajtowns commented:
> Why not put all arguments in `CNodeOptions`? I

It's only the optional arguments that are in there, which I think is a pretty reasonable dividing line, and also limits the code changes needed.

vasild commented on src/net.cpp:
```diff
@@ -1019,6 +1021,9 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
 
     const bool inbound_onion = std::find(m_onion_binds.begin(), m_onion_binds.end(), addr_bind) != m_onion_binds.end();
+    CNodeOptions node_opts;
+    node_opts.prefer_evict = discouraged;
```
`prefer_evict` is set here, but later ignored by the `CNode` constructor. Maybe this was forgotten:

```diff
--- i/src/net.cpp
+++ w/src/net.cpp
@@ -2733,12 +2733,13 @@ CNode::CNode(NodeId idIn,
       m_sock{sock},
       m_connected{GetTime<std::chrono::seconds>()},
       addr{addrIn},
       addrBind{addrBindIn},
       m_addr_name{addrNameIn.empty() ? addr.ToStringIPPort() : addrNameIn},
       m_inbound_onion{inbound_onion},
+      m_prefer_evict{node_opts.prefer_evict},
       nKeyedNetGroup{nKeyedNetGroupIn},
       id{idIn},
       nLocalHostNonce{nLocalHostNonceIn},
       m_conn_type{conn_type_in},
       m_i2p_sam_session{std::move(node_opts.i2p_sam_session)}
 {
```

vasild commented on src/net.cpp:
```diff
@@ -1019,6 +1021,9 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
     const bool inbound_onion = std::find(m_onion_binds.begin(), m_onion_binds.end(), addr_bind) != m_onion_binds.end();
+    CNodeOptions node_opts;
+    node_opts.prefer_evict = discouraged;
+    node_opts.permission_flags = permissionFlags;
```
It could be a bit shorter if designated initializers are used, instead of:

```cpp
Opts opts;
opts.x = 1;
opts.y = 2;
opts.z = 3;
new CNode(
    ...
    opts
);
```

it would become:

```cpp
new CNode(
    ...
    {
        .x = 1,
        .y = 2,
        .z = 3
    }
);
```

ajtowns replied:
Err, yeah; I had the same thought, but apparently didn't push it.

vasild reviewed (commented):
> It's only the optional arguments that are in there, which I think is a pretty reasonable dividing line, and also limits the code changes needed.

I see. Maybe it makes sense to put all arguments in `CNodeOptions`, but even as it is this PR seems like an improvement over `master`. Would be happy to review also if all arguments are passed via `CNodeOptions`.

vasild commented on src/net.h:
```diff
@@ -334,6 +334,13 @@ class V1TransportSerializer : public TransportSerializer {
     void prepareForTransport(CSerializedNetMsg& msg, std::vector<unsigned char>& header) const override;
 };
 
+struct CNodeOptions
```
nit: (feel free to ignore) If this is defined inside `CNode`, then callers would use `CNode::Options` instead of `CNodeOptions` and inside the class it would be just `Options` instead of `CNodeOptions`. Also, if/when `CNode` is renamed to something else, it would not be necessary to rename `CNodeOptions` too.

ajtowns replied:
Unfortunately there's a gcc/clang bug that means this doesn't work right:

```c++
class X
{
public:
    struct Y
    {
        int foo = 42;
    };

    int SomeFunc(Y y = {});
};
```

gives:

```
$ g++ -std=c++17 -Wall -W -c -o lame.o lame.cpp
lame.cpp:11:25: error: could not convert ‘<brace-enclosed initializer list>()’ from ‘<brace-enclosed initializer list>’ to ‘X::Y’
   11 |     int SomeFunc(Y y = {});
      |                         ^
      |                         |
      |                         <brace-enclosed initializer list>

$ clang -std=c++17 -Wall -W -c -o lame.o lame.cpp
lame.cpp:11:25: error: default member initializer for 'foo' needed within definition of enclosing class 'X' outside of member functions
    int SomeFunc(Y y = {});
                        ^
lame.cpp:7:13: note: default member initializer declared here
        int foo = 42;
            ^
1 error generated.
```

Moving `Y` outside of `X` fixes this. https://stackoverflow.com/a/53423881/ has more details.

vasild replied:
I see. Omitting ` = {}` from `SomeFunc()` definition seems to resolve this, but then callers need to pass `{}` at the end if they don't wish to pass optional arguments. Or if all arguments are passed via the Options variable, then it will be ok.

naumenkogs commented:
Concept ACK. Will do code review once the pending feedback is resolved.

ajtowns committed:
net: make CNode::m_permissionFlags const

ajtowns committed:
net: make CNode::m_prefer_evict const

ajtowns committed:
scripted-diff: net: rename permissionFlags to permission_flags

-BEGIN VERIFY SCRIPT-
sed -i 's/permissionFlags/permission_flags/g' $(git grep -l permissionFlags)
-END VERIFY SCRIPT-

vasild commented on src/test/fuzz/util.h:
```diff
@@ -310,7 +311,8 @@ auto ConsumeNode(FuzzedDataProvider& fuzzed_data_provider, const std::optional<N
                                        conn_type,
-                                       inbound_onion);
+                                       inbound_onion,
+                                       CNodeOptions{ .permission_flags = permission_flags });
```
nit: here and below, could omit `CNodeOptions`, like done in `net.cpp`:

```suggestion
                                       { .permission_flags = permission_flags });
```

vasild replied:
Or maybe not: CI is upset:

```
./net.h:523:5: note:   no known conversion for argument 10 from ‘<brace-enclosed initializer list>’ to ‘CNodeOptions&&’
```

ajtowns replied:
Can't omit it here because `std::make_unique<CNode>(...)` can't resolve to a specific function without knowing what type `{ .permission_flags = .. }` is going to be. (Could omit it in the `else` clause, but figured it was better to have those two clauses be as similar as possible)

ajtowns replied:
Looks like you can't omit it when using `.foo = std::move(unique_ptr)` either, as gcc 8 gives errors. https://godbolt.org/z/rbn6csbax

ryanofsky replied:
> Can't omit it here because `std::make_unique<CNode>(...)` can't resolve to a specific function without knowing what type `{ .permission_flags = .. }` is going to be.

This is ok, but an alternative is to switch from `make_unique<CNode>` to `new CNode`

vasild reviewed (approved):
ACK 2f3602b655cac8ada33ba2a5371c216ae1f1025e

jonatack commented:
Review ACK modulo the CI issue

jonatack reviewed (commented):
ACK 377e9ccda469731d535829f184b70c73ed46b6ef per `git range-diff 52dcb1d 2f3602b 377e9cc`

<details><summary>Some clang-format nits if you repush</summary><p>

```diff
diff --git a/src/net.cpp b/src/net.cpp
index 6659b64246..138486e468 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -556,7 +556,7 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
                              pszDest ? pszDest : "",
                              conn_type,
                              /*inbound_onion=*/false,
-                             CNodeOptions{ .i2p_sam_session = std::move(i2p_transient_session) });
+                             CNodeOptions{.i2p_sam_session = std::move(i2p_transient_session)});
     pnode->AddRef();

@@ -1029,8 +1027,8 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
                              ConnectionType::INBOUND,
                              inbound_onion,
                              CNodeOptions{
-                               .permission_flags = permission_flags,
-                               .prefer_evict = discouraged,
+                                 .permission_flags = permission_flags,
+                                 .prefer_evict = discouraged,
                              });
diff --git a/src/test/fuzz/util.h b/src/test/fuzz/util.h
index 6d652c922b..46cd422cf4 100644
--- a/src/test/fuzz/util.h
+++ b/src/test/fuzz/util.h
@@ -312,7 +312,7 @@ auto ConsumeNode(FuzzedDataProvider& fuzzed_data_provider, const std::optional<N
                                        addr_name,
                                        conn_type,
                                        inbound_onion,
-                                       CNodeOptions{ .permission_flags = permission_flags });
+                                       CNodeOptions{.permission_flags = permission_flags});
     } else {
         return CNode{node_id,
                      sock,
@@ -323,7 +323,7 @@ auto ConsumeNode(FuzzedDataProvider& fuzzed_data_provider, const std::optional<N
                      addr_name,
                      conn_type,
                      inbound_onion,
-                     CNodeOptions{ .permission_flags = permission_flags }};
+                     CNodeOptions{.permission_flags = permission_flags}};
     }
 }
```
</p></details>

vasild reviewed (approved):
ACK 377e9ccda469731d535829f184b70c73ed46b6ef

ryanofsky commented on src/net.cpp:
```diff
@@ -556,7 +556,7 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
                              conn_type,
                              /*inbound_onion=*/false,
-                             std::move(i2p_transient_session));
+                             CNodeOptions{ .i2p_sam_session = std::move(i2p_transient_session) });
```
In commit "net: add CNodeOptions for optional CNode constructor params" (9dccc3328eeaf9cd66518d812c878def5d014e36)

I'd shorten this as follows, but it's also reasonable to keep if you think the struct name is helpful.

```diff
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -556,7 +556,7 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
                              pszDest ? pszDest : "",
                              conn_type,
                              /*inbound_onion=*/false,
-                             CNodeOptions{ .i2p_sam_session = std::move(i2p_transient_session) });
+                             { .i2p_sam_session = std::move(i2p_transient_session) });
     pnode->AddRef();
 
     // We're making a new connection, harvest entropy from the time (and our peer count)

```

The short version is nice because it's basically a substitute for having named arguments in c++.

Same comment also applies to later commits, as well.

ajtowns replied:
Agree, but I don't think this is possible until we drop support for gcc 8? https://github.com/bitcoin/bitcoin/pull/25962#discussion_r960500526

ryanofsky reviewed (approved):
Code review ACK 377e9ccda469731d535829f184b70c73ed46b6ef. Looks good and feel free to ignore suggestions!

naumenkogs commented:
ACK 377e9ccda469731d535829f184b70c73ed46b6ef
